<a href="https://colab.research.google.com/github/OleksiiLatypov/DRU_Data_Science_Fundamentals/blob/main/test_task.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# How to complete this assignment
Complete 15 graded tasks below.

- Do not delete or rename the variables given before the inscription `#your code here`.
- Do not change the code in the last Notebook cell, it is required for the server check.

## Some important notes
- If you need to **calculate the number of "something"** that means we expect you to assign an Integer to the given variable
- If you need to **make a list of "something"** we expect you to assign a Python list with appropriate values to the given variable
- If you need to find a **specifiс player, day of the week, team, etc.** we expect you to assign a String with the full name of the entity to the given variable (`player_name`, day of week full name, `team_name`, etc.)

In [89]:
import pandas as pd

pd.set_option('display.max_column', None)

In [90]:
player_data = pd.read_csv("player_data.csv", index_col="Unnamed: 0")

In [92]:
player_data.head()

id  player_api_id         player_name  player_fifa_api_id  \
0   1         505942  Aaron Appindangoye              218353   
1   2         155782     Aaron Cresswell              189615   
2   3         162549         Aaron Doran              186170   
3   4          30572       Aaron Galindo              140161   
4   5          23780        Aaron Hughes               17725   

              birthday  height  weight  
0  1992-02-29 00:00:00  182.88     187  
1  1989-12-15 00:00:00  170.18     146  
2  1991-05-13 00:00:00  170.18     163  
3  1982-05-08 00:00:00  182.88     198  
4  1979-11-08 00:00:00  182.88     154

**Task 1 (0.25 point).** Calculate the number of players with a height between 180 and 190 inclusive

In [93]:
players_180_190 = player_data[player_data['height'].between(180, 190)]

**Task 2 (0.25 point).** Calculate the number of players born in 1980. <br>

In [94]:
players_1980 = player_data[player_data['birthday'].apply(lambda x: x[:4] == '1980')].shape[0]

**Task 3 (0.25 point).** Make a list of the top 10 players with the highest weight sorted in descending order. If there are several players with the same weight put them in the lexicographic order by name.

In [39]:
highest_players = player_data.sort_values(by=['weight', 'player_name'], ascending=[False, True])
highest_players

id  player_api_id       player_name  player_fifa_api_id  \
5901    5908         148325  Kristof van Hout              185306   
10297  10313          27313         Tim Wiese               53012   
4879    4884           5044  Jeroen Verhoeven              137692   
4408    4412          27267    Ishmael Miller              169725   
1595    1598         101584            Cassio              167958   
...      ...            ...               ...                 ...   
1477    1480         163840           Caetano              199931   
2609    2612         103139  Diego Buonanotte              170719   
6843    6853         306631   Mariusz Rybicki              208095   
10938  10954         108461   Yohandry Orozco              202559   
5446    5453          97226        Juan Quero              171126   

                  birthday  height  weight  
5901   1987-02-09 00:00:00  208.28     243  
10297  1981-12-17 00:00:00  193.04     243  
4879   1980-04-30 00:00:00  198.12     227  
4408   1987-03-05 00:00:00  193.04     225  
1595   1987-06-06 00:00:00  195.58     220  
...                    ...     ...     ...  
1477   1991-04-20 00:00:00  165.10     123  
2609   1988-04-19 00:00:00  160.02     123  
6843   1993-03-13 00:00:00  175.26     121  
10938  1991-03-19 00:00:00  165.10     121  
5446   1984-10-17 00:00:00  157.48     117  

[11060 rows x 7 columns]

**Task 4 (0.5 point).** Make a list of tuples containing years along with the number of players born in that year from 1980 up to 1990. <br>
**Structure example**: [(1980, 123), (1981, 140) ..., (1990, 83)] -> There were born 123 players in 1980, there were born 140 players in 1981 and etc.

In [88]:
years_df = pd.to_datetime(player_data['birthday'], errors='coerce')
years_df_1980_1990 = player_data[player_data['birthday'].between('1980', '1991', inclusive=True)]
#years_born_players = # Your code here
years_df_1980_1990.sort_values(by='birthday', ascending=False).head()
players_count_by_year = years_df_1980_1990.groupby('birthday').size().reset_index(name='count')
years_and_players = list(zip(players_count_by_year['birthday'], players_count_by_year['count']))
players_count_by_year
years_and_players

<ipython-input-88-9a60b1986611>:2: FutureWarning: Boolean inputs to the `inclusive` argument are deprecated in favour of `both` or `neither`.
  years_df_1980_1990 = player_data[player_data['birthday'].between('1980', '1991', inclusive=True)]


[('1980-01-01 00:00:00', 1),
 ('1980-01-03 00:00:00', 1),
 ('1980-01-04 00:00:00', 1),
 ('1980-01-05 00:00:00', 2),
 ('1980-01-06 00:00:00', 1),
 ('1980-01-08 00:00:00', 1),
 ('1980-01-09 00:00:00', 2),
 ('1980-01-11 00:00:00', 3),
 ('1980-01-16 00:00:00', 1),
 ('1980-01-18 00:00:00', 3),
 ('1980-01-19 00:00:00', 1),
 ('1980-01-21 00:00:00', 3),
 ('1980-01-22 00:00:00', 2),
 ('1980-01-25 00:00:00', 4),
 ('1980-01-26 00:00:00', 1),
 ('1980-01-27 00:00:00', 1),
 ('1980-01-29 00:00:00', 3),
 ('1980-01-30 00:00:00', 2),
 ('1980-01-31 00:00:00', 2),
 ('1980-02-01 00:00:00', 3),
 ('1980-02-02 00:00:00', 2),
 ('1980-02-03 00:00:00', 1),
 ('1980-02-06 00:00:00', 1),
 ('1980-02-07 00:00:00', 1),
 ('1980-02-08 00:00:00', 1),
 ('1980-02-09 00:00:00', 2),
 ('1980-02-10 00:00:00', 3),
 ('1980-02-11 00:00:00', 1),
 ('1980-02-12 00:00:00', 2),
 ('1980-02-13 00:00:00', 1),
 ('1980-02-14 00:00:00', 1),
 ('1980-02-17 00:00:00', 2),
 ('1980-02-18 00:00:00', 2),
 ('1980-02-20 00:00:00', 2),
 ('1980-02-21 

**Task 5 (0.5 point).** Calculate the mean and the standard deviation of the players' **height** with the name **Adriano**. <br>
**Note:** Name is represented by the first part of `player_name`.

In [ ]:
adriano_mean, adriano_std = # Your code here

**Task 6 (0.75 point).** How many players were born on each day of the week? Find the day of the week with the minimum number of players born.

In [ ]:
dow_with_min_players_born = # Your code here

**Task 7 (0.75 point).** Find a league with the most matches in total. If there are several leagues with the same amount of matches, take the first in the lexical order.

*Note*: Use league_data.csv and match_data.csv

In [ ]:
league_data = pd.read_csv("league_data.csv", index_col="Unnamed: 0")
match_data = pd.read_csv("match_data.csv", index_col="Unnamed: 0")

In [ ]:
league_data.head()

league_id  country_id             league_name
0          1           1  Belgium Jupiler League
1       1729        1729  England Premier League
2       4769        4769          France Ligue 1
3       7809        7809   Germany 1. Bundesliga
4      10257       10257           Italy Serie A

In [ ]:
match_data.head()

id  country_id  league_id     season  stage                 date   
0   1           1          1  2008/2009      1  2008-08-17 00:00:00  \
1   2           1          1  2008/2009      1  2008-08-16 00:00:00   
2   3           1          1  2008/2009      1  2008-08-16 00:00:00   
3   4           1          1  2008/2009      1  2008-08-17 00:00:00   
4   5           1          1  2008/2009      1  2008-08-16 00:00:00   

   match_api_id  home_team_api_id  away_team_api_id  home_team_goal   
0        492473              9987              9993               1  \
1        492474             10000              9994               0   
2        492475              9984              8635               0   
3        492476              9991              9998               5   
4        492477              7947              9985               1   

   away_team_goal  home_player_X1  home_player_X2  home_player_X3   
0               1             NaN             NaN             NaN  \
1               0             NaN             NaN             NaN   
2               3             NaN             NaN             NaN   
3               0             NaN             NaN             NaN   
4               3             NaN             NaN             NaN   

   home_player_X4  home_player_X5  home_player_X6  home_player_X7   
0             NaN             NaN             NaN             NaN  \
1             NaN             NaN             NaN             NaN   
2             NaN             NaN             NaN             NaN   
3             NaN             NaN             NaN             NaN   
4             NaN             NaN             NaN             NaN   

   home_player_X8  home_player_X9  home_player_X10  home_player_X11   
0             NaN             NaN              NaN              NaN  \
1             NaN             NaN              NaN              NaN   
2             NaN             NaN              NaN              NaN   
3             NaN             NaN              NaN              NaN   
4             NaN             NaN              NaN              NaN   

   away_player_X1  away_player_X2  away_player_X3  away_player_X4   
0             NaN             NaN             NaN             NaN  \
1             NaN             NaN             NaN             NaN   
2             NaN             NaN             NaN             NaN   
3             NaN             NaN             NaN             NaN   
4             NaN             NaN             NaN             NaN   

   away_player_X5  away_player_X6  away_player_X7  away_player_X8   
0             NaN             NaN             NaN             NaN  \
1             NaN             NaN             NaN             NaN   
2             NaN             NaN             NaN             NaN   
3             NaN             NaN             NaN             NaN   
4             NaN             NaN             NaN             NaN   

   away_player_X9  away_player_X10  away_player_X11  home_player_Y1   
0             NaN              NaN              NaN             NaN  \
1             NaN              NaN              NaN             NaN   
2             NaN              NaN              NaN             NaN   
3             NaN              NaN              NaN             NaN   
4             NaN              NaN              NaN             NaN   

   home_player_Y2  home_player_Y3  home_player_Y4  home_player_Y5   
0             NaN             NaN             NaN             NaN  \
1             NaN             NaN             NaN             NaN   
2             NaN             NaN             NaN             NaN   
3             NaN             NaN             NaN             NaN   
4             NaN             NaN             NaN             NaN   

   home_player_Y6  home_player_Y7  home_player_Y8  home_player_Y9   
0             NaN             NaN             NaN             NaN  \
1             NaN             NaN             NaN             NaN   
2            

In [ ]:
league_most_matches = # Your code here

**Task 8 (1.25 point).** Find a player who participated in the largest number of matches during the whole match history. Assign a `player_name` to the given variable (use match_data['home_player_N', 'away_player_N'])

In [ ]:
max_matches_player = # Your code here

**Task 9 (1.5 point).** List top-5 tuples of most correlated **player's characteristics** in the descending order of the absolute [Pearson's coefficient](https://en.wikipedia.org/wiki/Pearson_correlation_coefficient) value.

**Note 1:** Players characteristics are all the columns in `Player_Attributes` table except `[id, player_fifa_api_id, player_api_id, date, preferred_foot, attacking_work_rate, defensive_work_rate]`). <br>
**Note 2:** Exclude duplicated pairs from the list. E.g. ('gk_handling', 'gk_reflexes') and ('gk_reflexes', 'gk_handling') are duplicates, leave just one of them in the resulting list.

In [ ]:
player_attributes = pd.read_csv('player_attributes.csv', index_col="Unnamed: 0")

In [ ]:
player_attributes.head()

id  player_fifa_api_id  player_api_id                 date  overall_rating   
0   1              218353         505942  2016-02-18 00:00:00            67.0  \
1   2              218353         505942  2015-11-19 00:00:00            67.0   
2   3              218353         505942  2015-09-21 00:00:00            62.0   
3   4              218353         505942  2015-03-20 00:00:00            61.0   
4   5              218353         505942  2007-02-22 00:00:00            61.0   

   potential preferred_foot attacking_work_rate defensive_work_rate  crossing   
0       71.0          right              medium              medium      49.0  \
1       71.0          right              medium              medium      49.0   
2       66.0          right              medium              medium      49.0   
3       65.0          right              medium              medium      48.0   
4       65.0          right              medium              medium      48.0   

   finishing  heading_accuracy  short_passing  volleys  dribbling  curve   
0       44.0              71.0           61.0     44.0       51.0   45.0  \
1       44.0              71.0           61.0     44.0       51.0   45.0   
2       44.0              71.0           61.0     44.0       51.0   45.0   
3       43.0              70.0           60.0     43.0       50.0   44.0   
4       43.0              70.0           60.0     43.0       50.0   44.0   

   free_kick_accuracy  long_passing  ball_control  acceleration  sprint_speed   
0                39.0          64.0          49.0          60.0          64.0  \
1                39.0          64.0          49.0          60.0          64.0   
2                39.0          64.0          49.0          60.0          64.0   
3                38.0          63.0          48.0          60.0          64.0   
4                38.0          63.0          48.0          60.0          64.0   

   agility  reactions  balance  shot_power  jumping  stamina  strength   
0     59.0       47.0     65.0        55.0     58.0     54.0      76.0  \
1     59.0       47.0     65.0        55.0     58.0     54.0      76.0   
2     59.0       47.0     65.0        55.0     58.0     54.0      76.0   
3     59.0       46.0     65.0        54.0     58.0     54.0      76.0   
4     59.0       46.0     65.0        54.0     58.0     54.0      76.0   

   long_shots  aggression  interceptions  positioning  vision  penalties   
0        35.0        71.0           70.0         45.0    54.0       48.0  \
1        35.0        71.0           70.0         45.0    54.0       48.0   
2        35.0        63.0           41.0         45.0    54.0       48.0   
3        34.0        62.0           40.0         44.0    53.0       47.0   
4        34.0        62.0           40.0         44.0    53.0       47.0   

   marking  standing_tackle  sliding_tackle  gk_diving  gk_handling   
0     65.0             69.0            69.0        6.0         11.0  \
1     65.0             69.0            69.0        6.0         11.0   
2     65.0             66.0            69.0        6.0         11.0   
3     62.0             63.0            66.0        5.0         10.0   
4     62.0             63.0            66.0        5.0         10.0   

   gk_kicking  gk_positioning  gk_reflexes  
0        10.0             8.0          8.0  
1        10.0             8.0          8.0  
2        10.0             8.0          8.0  
3         9.0             7.0          7.0  
4         9.0             7.0          7.0

In [ ]:
top_correlated_features = # Your code here

**Task 10 (2 points).** Find top-5 most similar players to **Neymar** whose names are given. The similarity is measured as [Euclidean distance](https://en.wikipedia.org/wiki/Euclidean_distance) between vectors of players' characteristics (described in the task above). Put their names in a vector in ascending order by Euclidean distance and sorted by `player_name` if the distance is the same<br>
**Note 1:** There are many records for some players in the `Player_Attributes` table. You need to take the freshest data (characteristics with the most recent `date`). <br>
**Note 2:** Use pure values of the characteristics even if you are aware of such preprocessing technics as normalization. <br>
**Note 3:** Please avoid using any built-in methods for calculating the Euclidean distance between vectors, think about implementing your own.

In [ ]:
neymar_similarities = # Your code here

**Task 11 (1 point).** Calculate the number of home matches played by the **Borussia Dortmund** team in **Germany 1. Bundesliga** in season **2008/2009**

In [ ]:
team_data = pd.read_csv('team_data.csv', index_col="Unnamed: 0")

In [ ]:
team_data.head()

id  team_api_id  team_fifa_api_id     team_long_name team_short_name
0   1         9987             673.0           KRC Genk             GEN
1   2         9993             675.0       Beerschot AC             BAC
2   3        10000           15005.0   SV Zulte-Waregem             ZUL
3   4         9994            2007.0   Sporting Lokeren             LOK
4   5         9984            1750.0  KSV Cercle Brugge             CEB

In [ ]:
borussia_bundesliga_2008_2009_matches = # Your code here

**Task 12 (1 point).** Find a team having the most matches (both home and away!) in the **Germany 1. Bundesliga** in **2008/2009** season. Return number of matches.

In [ ]:
team_most_matches_bundesliga_2008_2009 = # Your code here

**Task 13 (1 point).** Count total number of **Arsenal** matches (both home and away!) in the **2015/2016** season which they have won. <br><br>
**Note:** Winning a game means scoring **more** goals than an opponent.

In [ ]:
arsenal_won_matches_2015_2016 = # Your code here

**Task 14 (2 points).** Find a team with the highest win rate in the **2015/2016** season. Win rate means won matches / all matches. If there are several teams with the highest win rate return the first by name in lexical order

In [ ]:
team_highest_winrate_2015_2016 = # Your code here

**Task 15 (2 points).** Determine the team with the maximum days' gap between matches in **England Premier League 2010/2011 season**. Return number of days in that gap. <br>
**Note**: a *gap* means the number of days between two consecutive matches of the same team.

In [ ]:
highest_gap_england_2010_2011 = # Your code here

### Warning! Do not change anything in the area below

In [ ]:
with open('student_answers.txt', 'w') as file:
    file.write(f"{players_180_190}\n")
    file.write(f"{players_1980}\n")
    file.write(f"{highest_players}\n")
    file.write(f"{years_born_players}\n")
    file.write(f"{round(adriano_mean, 3)} {round(adriano_std, 3)}\n")
    file.write(f"{dow_with_min_players_born}\n")
    file.write(f"{league_most_matches}\n")
    file.write(f"{max_matches_player}\n")
    file.write(f"{';'.join(['%s,%s' % tup for tup in top_correlated_features])};\n")
    file.write(f"{neymar_similarities}\n")
    file.write(f"{borussia_bundesliga_2008_2009_matches}\n")
    file.write(f"{team_most_matches_bundesliga_2008_2009}\n")
    file.write(f"{arsenal_won_matches_2015_2016}\n")
    file.write(f"{team_highest_winrate_2015_2016}\n")
    file.write(f"{highest_gap_england_2010_2011}\n")